viamonvif/service.go

```diff
@@ -175,21 +178,43 @@ func (dis *rtspDiscovery) DoCommand(ctx context.Context, command map[string]inte
 		if err != nil {
 			return nil, err
 		}
+
+		var snapshotErr, rtspErr error
+
+		// Attempt to fetch the image from the snapshot URI
 		dis.rtspToSnapshotURIsMu.Lock()
 		snapshotURI, found := dis.rtspToSnapshotURIs[previewReq.rtspURL]
 		dis.rtspToSnapshotURIsMu.Unlock()
-		if !found {
-			return nil, fmt.Errorf("snapshot URI not found for %s", previewReq.rtspURL)
+
+		if found {
+			dis.logger.Debugf("attempting to fetch image from snapshot URI: %s", snapshotURI)
+			dataURL, err := downloadPreviewImage(ctx, dis.logger, snapshotURI)
+			if err == nil {
+				dis.logger.Debugf("Successfully fetched image from snapshot URI: %s", snapshotURI)
+				return map[string]interface{}{
+					"preview": dataURL,
+				}, nil
+			}
+			dis.logger.Warnf("failed to fetch image from snapshot URI: %s, error: %v", snapshotURI, err)
+			snapshotErr = fmt.Errorf("snapshot error: %w", err)
+		} else {
+			dis.logger.Warnf("snapshot URI not found for RTSP URL: %s", previewReq.rtspURL)
+			snapshotErr = errors.New("snapshot URI not found")
 		}
-		dis.logger.Infof("snapshot URI: %s", snapshotURI)
 
-		dataURL, err := downloadPreviewImage(ctx, dis.logger, snapshotURI)
-		if err != nil {
-			return nil, fmt.Errorf("failed to download preview image: %w", err)
+		// Fallback to fetching the image via RTSP
+		dis.logger.Debugf("attempting to fetch image via RTSP URL: %s", previewReq.rtspURL)
+		dataURL, err := fetchImageFromRTSPURL(ctx, dis.logger, previewReq.rtspURL)
+		if err == nil {
+			dis.logger.Debugf("successfully fetched image via RTSP for URL: %s", previewReq.rtspURL)
+			return map[string]interface{}{
+				"preview": dataURL,
+			}, nil
 		}
-		return map[string]interface{}{
-			"preview": dataURL,
-		}, nil
+		dis.logger.Warnf("failed to fetch image via RTSP for URL: %s, error: %v", previewReq.rtspURL, err)
+		rtspErr = fmt.Errorf("RTSP error: %w", err)
+
+		return nil, fmt.Errorf("both snapshot and RTSP fetch failed: %w; %w", snapshotErr, rtspErr)
```

Comment:
is `errors.Join()` better here than two %w formatters

Reply:
Like that, changed.